### PR TITLE
Fixes for 3.13

### PR DESCRIPTION
**Pull request by A5rocks**
Fixes https://github.com/python-trio/trio/issues/3004, also refs https://github.com/python-trio/trio/issues/2885

@njsmith could you check my naive `locals()` -> `sys._getframe().f_locals` replacement? Unfortunately 3.13 makes `locals()` mutations not mutate actual local variables, as part of PEP 667. I'm especially not sure about the replacement in `_run.py` where `locals()` used to clear something (not sure what).

### CODE REVIEW

File: src/trio/_core/_run.py
Comment (hroncok) [reply src/trio/_core/_run.py:664]: Perhaps something like:

```python
if sys.version_info < (3, 13):
    locals()
```

**Comment by richardsheridan** [comment]
Seems like with these changes `CancelScope.__exit__` doesn't need to inline the KI protection decorator anymore, so I pushed a commit on that.

It seems like this PR is done, are you ready for someone to merge yet?

File: src/trio/_core/_run.py
Comment (A5rocks) [reply src/trio/_core/_run.py:664]: Just to confirm: latest state of the PR fails tests on fedora? (That's not a platform in CI so very possible)

Comment (oremanj) [reply src/trio/_core/_run.py:664]: I would expect you to still need the dummy call to `locals()` on 3.12 and earlier. It looks like accessing either `locals()` or `f_locals` behaves the same way pre 3.13: it creates a dictionary with the current contents of the fast locals, but the dictionary is not updated on later changes (unless a trace function is set) until you access again.

You've removed the call to `locals()` above so if that's correct then you probably don't need the 2nd call here. I would switch back to the previous version (inlining `enable_ki_protection`, but update it to write to `sys._getframe().f_locals`) and keep the redundant call to `locals()` but only on 3.12 and earlier.

Comment (befeleme) [reply src/trio/_core/_run.py:664]: > Just to confirm: latest state of the PR fails tests on fedora with Python 3.12? (That's not a platform in CI so very possible)

Yes, we're testing on Fedora

Comment (A5rocks) [reply src/trio/_core/_run.py:664]: I'm a bit confused cause when I used Python 3.12.3 on Fedora (Docker, I built CPython 3.12.3 from source) and `pytest --pyargs trio --skip-optional-imports` on this PR, the tests pass. Am I missing something?

Comment (hroncok) [reply src/trio/_core/_run.py:664]: As said, we are stuck at 0.23.1, we are backporting this PR on top of it, so perhaps that makes a difference.

Comment (jakkdl) [reply src/trio/_core/_run.py:664]: > As said, we are stuck at 0.23.1, we are backporting this PR on top of it, so perhaps that makes a difference.

what's the reason you're stuck at 0.23.1? Looking at the changelog of 0.23.2 doesn't look like it should be introducing anything controversial.

Comment (hroncok) [reply src/trio/_core/_run.py:664]: https://github.com/python-trio/trio-websocket/issues/187 and possibly others

Comment (jakkdl) [reply src/trio/_core/_run.py:664]: that should only be affected from 0.25.0 though. If you're at the point where you're cherrypicking commits you could instead revert https://github.com/python-trio/trio/commit/cfc0755e8ef1d9d7e05de4f39d6c2bdbad02fdf0

**Comment by A5rocks** [comment]
I think this is ready for review/merge. While this applied as a patch doesn't seem to work, I can't seem to reproduce that and I'm sure it's just some other PR that changed things just enough for this subtle locals stuff to work (which can be located probably just using a `git bisect`?).

**Comment by jakkdl** [comment]
I think you can get 3.13 to pass in CI if you explicitly install `cffi==1.17.0rc1` as a dependency: https://github.com/python-cffi/cffi/releases/tag/v1.17.0rc1
Since it's a release candidate pip doesn't automatically target it.


I wrote a PR to get trio-websocket to support trio>0.25 https://github.com/python-trio/trio-websocket/pull/188, so once merged that may fully resolve the issues encountered by the distro maintainers.

**Comment by A5rocks** [comment]
Alright, this is now running the whole test suite (rather than the part that doesn't require the test requirements).

**Review by CoolCat467 — APPROVED** [review]
Makes sense, looks good as long as others are in agreement that the locals magic thing discussion is taken care of.

**Review by jakkdl — APPROVED** [review]
I also don't understand `locals()` magic, but the tests are passing and the other changes looks good.

**Review by richardsheridan — APPROVED** [review]
The locals()/garbage interaction isn't triggering some of the tests the way it was when I inserted it. I don't think it's worth bisecting python/trio to understand it or preemptively fixing, since it will be at worst a performance regression. If someone eventually stumbles on a reproducible regression to add to the tests, we know where to look for the fix.

**Pull request merged**
